File: ruby/anagram/anagram.rb
class Anagram
  def initialize(word)
    @word = word
  end

  def match(possible_anagrams)
    possible_anagrams.find_all do |anagram|
      permutations.any? do |permutation|
        anagram.downcase == permutation
      end
    end
  end

  def permutations
    @word.downcase.chars.permutation.map(&:join) - [@word]
  end
end

File: javascript/anagram/anagram.js
function Anagram(subject){
    "use strict";

    this.match = function(possibleAnagrams) {
        return possibleAnagrams.filter(function(currentAnagram){
            return (areAnagramFriends(currentAnagram, subject));
        });
    };

    var areAnagramFriends = function(a, b) {
        return (a != b) && (normalizeString(a) == normalizeString(b));
    };

    var normalizeString = function(string) {
        return string.toLowerCase().split('').sort().join('');
    };
}

module.exports = Anagram;
